System locale

ashu26jha opened the pull request:
Mentors: @robertbastian @zbraniecki 

Total tasks:

- [x] Retrieval of system locale for MacOS, Windows, Linux
- [x] Calendar preferences
- [x] Time Zone for macos and windows
- [ ] Convert `String` into `icu4x` components

Optional: 

- [ ] Number preferences
- [ ] Currency preferences
- [ ] Temperature
- [ ] First Day of week

ashu26jha commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+            if !locale_ptr.is_null() {
+                let c_str = CStr::from_ptr(locale_ptr);
+                if let Ok(str_slice) = c_str.to_str() {
+                    return vec![(str_slice.to_string(), String::from("Gregorian"))];
```
Gregorian is default calendar for linux (doesn't change even changing the locale, won't work unless install extensions). Ubuntu uses GNOME calendar. Here is something in their repo: https://gitlab.gnome.org/GNOME/gnome-calendar/-/issues/998

robertbastian replied:
please add this as a code comment

robertbastian replied:
still needs to be done

robertbastian commented on utils/system_locale/src/lib.rs:
```diff
@@ -0,0 +1,30 @@
+#[cfg(target_os = "windows")]
+mod windows;
+
+pub fn get_locale() -> Vec<String> {
+    #[cfg(target_os = "linux")]
+    return linux::get_locale();
+
+    #[cfg(target_os = "macos")]
+    return apple::get_locale();
+
+    #[cfg(target_os = "windows")]
+    return windows::get_locale();
+}
+
+pub fn get_system_calendar() -> Vec<(String, String)> {
+    #[cfg(target_os = "linux")]
+    return linux::get_system_calendar();
+
+    #[cfg(target_os = "macos")]
+    return apple::get_system_calendar();
+
+    #[cfg(target_os = "windows")]
+    return windows::get_system_calendar();
+}
```
reexport the methods instead of wrapping them:

```suggestion
#[cfg(target_os = "linux")]
pub use linux::*;

#[cfg(target_os = "macos")]
pub use apple::*;

#[cfg(target_os = "windows")]
pub use windows::*;
```

robertbastian replied:
still needs to be done?

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+#[cfg(target_os = "linux")]
+mod linux_locale {
+    use locale::{
```
I don't think we want to depend on the `locale` crate, we're kind of the rewrite it wants to do: https://github.com/rust-locale/rust-locale/issues/7 (this crate is the `langinfo_sys` part). @zbraniecki do we want to revive the discussions in `rust-locale` at this time?

`set_locale` is available in [`libc`](https://docs.rs/libc/0.2.155/libc/fn.setlocale.html), and the constants you should define yourself.

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+        let mut locale_map = HashMap::new();
+
+        if let Ok(str_slice) = c_str.to_str() {
+            if str_slice.contains(';') {
```
`contains` requires a full scan of the string, try to write the algorithm with just `split`.

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+        ptr::null,
+    };
+
+    pub unsafe fn fetch_locale_settings() -> HashMap<i32, String> {
```
this method should be either be safe, or private. If safe, any unsafeness should be encapsulated inside. If unsafe, it should have a comment stating the requirements for it to be safe.

zbraniecki replied:
I'd suggest abstracting the `LC_*` to an enum and have this return `HashMap<THIS_ENUM, Locale>`

robertbastian replied:
agreed

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+    use std::{
+        collections::{HashMap, HashSet},
+        ffi::CStr,
+        ptr::null,
```
nit: don't import free functions. Import `ptr` and call it as `ptr::null()`

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    pub fn get_system_calendar() -> Vec<(String, String)> {
+        unsafe {
```
use one `unsafe` block per `unsafe` function call, and explain the safety of each block in a comment

robertbastian replied:
still needs to be done everywhere

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+        unique_locales.into_iter().collect()
+    }
+
+    pub fn get_system_calendar() -> Vec<(String, String)> {
```
Try to reduce allocations by using this signature:

```suggestion
    pub fn get_system_calendar() -> impl Iterator<Item = (String, Cow<'static, str>)> {
```

You can use the `Option` type to return an 0-or-1-iterator:

```rust
if foo {
    Some(bar).into_iter()
} else {
    None.into_iter()
}
```

For a compile-time constant like `"Gregorian"` we don't need to allocate a `String`, and can instead use `Cow::Borrowed("Gregorian")`.

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+        locale_map
+    }
+
+    pub fn get_locale() -> Vec<String> {
+        let mut unique_locales = HashSet::new();
+        unsafe {
+            let locale_map = fetch_locale_settings();
+            for value in locale_map.values() {
+                unique_locales.insert(value.clone());
+            }
+        }
+        unique_locales.into_iter().collect()
+    }
```
See comments below.

```suggestion
    pub fn get_locale() -> impl Iterator<Item = String> {
        // SAFETY: <explain why this call is safe>
        unsafe { fetch_locale_settings() }.into_values()
    }
```

robertbastian replied:
still needs to be done, now it's just `fetch_locale_settings().into_values()`

robertbastian commented on utils/system_locale/src/lib.rs:
```diff
@@ -0,0 +1,30 @@
+#[cfg(target_os = "windows")]
+mod windows;
+
+pub fn get_locale() -> Vec<String> {
```
This returns a `Vec`, should it be `get_locales`?

robertbastian replied:
still needs to be done for windows

robertbastian commented on utils/system_locale/src/lib.rs:
```diff
@@ -0,0 +1,30 @@
+    return windows::get_locale();
+}
+
+pub fn get_system_calendar() -> Vec<(String, String)> {
```
also, plural?

zbraniecki replied:
`get_env_calendars` ?

robertbastian replied:
still needs to be done for windows

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+#[cfg(target_os = "linux")]
+mod linux_locale {
```
This file (= module) is already gated on linux, you don't need a nested module.
```suggestion
```

robertbastian replied:
still nested

robertbastian commented on utils/system_locale/src/linux.rs:
```diff
@@ -0,0 +1,79 @@
+        }
+        Vec::new()
+    }
+}
+
+#[cfg(target_os = "linux")]
+pub use linux_locale::get_locale;
+pub use linux_locale::get_system_calendar;
```
```suggestion
```

robertbastian commented on utils/system_locale/src/apple.rs:
please apply the same changes as in linux.rs

robertbastian commented on utils/system_locale/src/windows.rs:
please apply the same changes as in linux.rs

zbraniecki commented on Cargo.toml:
```diff
@@ -66,6 +66,7 @@ members = [
     "utils/litemap",
     "utils/pattern",
     "utils/resb",
+    "utils/system_locale",
```
This name suggests that the crate only retrieves "system" "locale". My understanding is that your crate aims to retrieve environment information from the runtime. I'd use a term like "environment", "preferences" etc.

My concern about the "system" is that it indicates that what you get is "OS Locale" - but you may not. Increasingly OSes sandbox applications, so OS may be in English, but the application may only see that the language of its own environment is French.
My concern about "locale" is that it indicates that all you get is a Unicode Locale - while what you actually get are things beyond that. For example, short date pattern is not part of a Locale but should be retrievable by this crate.

Maybe "utils/env_preferences"?

robertbastian commented on utils/env_preferences/src/linux.rs:
```diff
@@ -2,70 +2,118 @@
-    fn fetch_locale_settings() -> HashMap<String, String> {
+    #[derive(Hash, Eq, PartialEq, Debug)]
+    pub enum LocaleCategory {
+        LcCTYPE,
```
issue: these don't have to match the C names. You can give them ergonomic names like `Numeric`.

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -42,11 +44,13 @@ pub mod apple_locales_prefs {
         let mut calendar_identifier_str = String::new();
 
+        // SAFETY: The call to `CFCalendarCopyCurrent` returns a calendar object owned by us
+        // This calendar object is used extract locale and type of calendar (identifier)
```
issue: use one safety block and safety comment per unsafe call

robertbastian commented on utils/env_preferences/src/linux.rs:
```diff
@@ -2,70 +2,118 @@
-        // Thread safety is ensured by fallbacking to the default locale of `linux` which is `C`
+        // SAFETY: In case `libc::setlocale` returns a NULL pointer it fallbacks to the default locale "C"
+
         unsafe {
```
issue: use one unsafe block and unsafe comment per unsafe call

robertbastian commented on utils/env_preferences/src/linux.rs:
```diff
@@ -2,70 +2,118 @@
+                return locale_map;
+            }
+
+            // SAFETY: Creating a `CStr` from a non-null pointer and no mutation is being performed.
```
`Cstr::from_ptr` has more safety invariants than the pointer being non-null: https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_ptr. I at least want to see the null-termination mentioned.

robertbastian commented on utils/env_preferences/src/linux.rs:
```diff
@@ -2,70 +2,118 @@
+
+            let locales_cstr = CStr::from_ptr(locales_ptr);
+
+            // SAFETY: Returns `&[str]` slice
```
what does this refer to?

robertbastian commented on utils/env_preferences/src/linux.rs:
```diff
@@ -2,70 +2,118 @@
         unsafe {
             let locale_ptr = setlocale(LC_TIME, ptr::null());
+
+            // SAFETY: In case we get a `NULL` pointer for `LC_TIME` from `setlocale`, fallbacks
```
this doesn't seem to be a safety comment

ashu26jha commented:
1. Remove `cfg_if`
2. Added ergonomic names in `ENUM` 
3. Made sure there is one unsafe call per unsafe block
4. Added safety comments for apple.rs
5. Added safety comments for linux.rs

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,132 @@
+    pub fn get_locales() -> Vec<String> {
+        let mut languages: Vec<String> = Vec::new();
+
+        // SAFETY: The call to `CFLocaleCopyPreferredLanguages` returns an immutable reference to `CFArray` which is owned by us
```
if we own the `CFArray`, do we need to free it later?

ashu26jha replied:
Sure thing

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,132 @@
+
+        // SAFETY: The call to `CFCalendarCopyCurrent` returns a calendar object owned by us
+        // This calendar object is used extract locale and type of calendar (identifier)
+        let calendar = unsafe { CFCalendarCopyCurrent() };
```
same here

ashu26jha replied:
Released it with `CFRelease` 👍

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,132 @@
+            // SAFETY: Release the calendar when done to avoid memory leaks
+            unsafe { CFRelease(calendar as _) };
+
+            calendars.push((calendar_locale_str, calendar_identifier_str.clone()));
```
I think this `.clone()` is unnecessary?

robertbastian reviewed (commented):
Thanks for adding fine-grained safety comments. They look good overall, just one open comment about ownership on Mac.

robertbastian reviewed (dismissed):
LGTM, but would also like approval from @zbraniecki

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,136 @@
+                        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+                        let lang_rust_str = unsafe { CStr::from_ptr(lang_str) }
+                            .to_str()
+                            .unwrap_or("Unknown")
```
why would you return `Unknown` here? I think the best choice is to fail explicitly or silently - either return Error (make your function return Result<Vec<String>, YourError>) or silently - skip this if you can't acquire it.

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,136 @@
+
+        if !locale_carr_ref.is_null() {
+            // SAFETY: The call to `CFArrayGetCount` is only made when is `locale_carr_ref` is not `NULL`
+            let count = unsafe { CFArrayGetCount(locale_carr_ref as _) };
```
at this point you can be optimistic and call `locales.reserve(count)` to avoid reallocating in the loop.

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,136 @@
+        if languages.is_empty() {
+            languages.push(String::from("und"));
+        }
+        languages
```
I would not do this in this low level function. In my mental model:
* Low Level - get information from env. No logic. If languages is empty, return empty. If it fails, return failure.
* High level - normalize results from different platforms into a single cohesive result. Here comes heuristics such as this.

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,136 @@
+        languages
+    }
+
+    pub fn get_system_calendars() -> Vec<(String, String)> {
```
extract it as a named type. For example:

```
struct AppleCalendarMapping {
  locale: String,
  calendar: String
};
```

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,136 @@
+    pub fn get_system_calendars() -> Vec<(String, String)> {
+        let mut calendars = Vec::new();
+        let mut calendar_locale_str = String::new();
+        let mut calendar_identifier_str = String::new();
```
there's no value in defining those two strings here. Do it inside the `if !calendar.is_null()`. You can also do `let mut calendar_locale_str: Option<String> = None;

then try to get it.

Then check if you have `if let Some((locale, calendar)) = locale.zip(calendar) { calendars.push((...)) }`

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,136 @@
+
+        if calendars.is_empty() {
+            calendars.push((String::from("und"), String::from("Gregorian")));
+        }
```
again, I wouldn't change the values here. Let it return what the system has.

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,112 @@
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+pub mod linux_prefs {
```
you don't need a module here.

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,112 @@
+                "LC_IDENTIFICATION" => Ok(LocaleCategory::Identification),
+                "LC_ALL" => Ok(LocaleCategory::All),
+                _ => Ok(LocaleCategory::Other(s.to_string())),
+            }
```
`nit`: you can wrap your response in Ok and use `Self`:

```
Ok(match s {
  "LC_CTYPE" => Self::Character,
  "LC_NUMERIC" => Self::Number,
  ...
})
```

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,112 @@
+        let locales_ptr = unsafe { setlocale(LC_ALL, ptr::null()) };
+
+        if locales_ptr.is_null() {
+            locale_map.insert(LocaleCategory::All, "C".to_string());
```
low level - why are you adding a value here? If locale_ptr is null, locale_map should return null here.

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,112 @@
+            let locale_pairs = locales_str.split(';');
+
+            // To handle cases in case a single locale is returned or a list of locale
+            if locale_pairs.clone().count() == 1 {
```
no need to clone to get a count.

ashu26jha replied:
`count` takes ownership, so that is why I `clone` it

ashu26jha replied:
We loose ownership of locale_pairs when counting

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,112 @@
+
+            // To handle cases in case a single locale is returned or a list of locale
+            if locale_pairs.clone().count() == 1 {
+                locale_map.insert(LocaleCategory::All, "C".to_string());
```
Don't insert

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,112 @@
+        locale_map
+    }
+
+    pub fn get_system_calendars() -> impl Iterator<Item = (Cow<'static, str>, Cow<'static, str>)> {
```
why Cow? And Why Iterator? I'd return Vec<> of Strings. It looks like premature optimization.

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,112 @@
+                ))
+                .into_iter()
+                .chain(None);
+            }
```
What is `gnome-calendar` here? How does it work on non-gnome Linux? Why are you adding `Gregorian`?

I'd just return locale without calendar here if your API on Linux doesn't  have any info on calendars. just return `fn get_system_calendars() -> Result<Vec<String>>` for locales and document that all locales are for gregorian calendar.

ashu26jha replied:
I think the format should:

```rust
Result<String>
// LC_TIME -> only single locale can be returned
```

zbraniecki reviewed (changes requested):
One uber-comment - I think all of your per-platform functions should return Result. You can start with a single cross-platform Error enum.

Second thought - don't try to return the same type from per-platform. Each platform may have different API and capabilities. It's ok for Mac to return a list of strings, and Windows to return a Map of string. Return what the platform provides in the low level per-platform API.

In the high level cross-platform API take the per-platform and normalize it into a single ICU4X return value.

sffc reviewed (dismissed):
There's a lot of unsafe code in here (FFI calls). Because of the unsafe code, this cannot land in the ICU4X repository without thorough test coverage. I understand that you can't actually verify the return values since they are platform dependent, but you can test that they are all syntactically valid locales, as well as any other invariants the functions might have. Certainly at least ensure that absolutely every code path is covered and that its output is read and verified.

ashu26jha commented on utils/env_preferences/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+    "Globalization",
+    "Globalization_DateTimeFormatting",
+    "Win32",
+    "Win32_Globalization"
```
In window.rs, this is how they have defined the features, ref: 

mod.rs of windows crate
```rust
#[cfg(feature = "Globalization")]
pub mod Globalization;

#[cfg(feature = "System")]
pub mod System;
```

zbraniecki replied:
Oh, I'm so sorry. I misunderstood here. Those are not features you are introducing. You're using windows-rs features. Please disregard my comment. This is good.

zbraniecki commented on utils/env_preferences/src/windows.rs:
```diff
@@ -0,0 +1,37 @@
+                locale_vec_str.push(string);
+            }
+        }
+        Err(e) => return Err(RetrievalError::Other(e.to_string())),
```
This seems to be the only place where you use `Error::Other`. Generally `Other` is lossy - when as a customer of your API I see `Other` the way you defined it I can't reason much about it - I can only display it to the user.

But you don't need to make it lossy. You can add a variant called `Windows` to your error which embeds https://docs.rs/windows-result/0.2.0/windows_result/struct.Error.html which is an Error returned by all Windows.rs crates.

```
pub enum RetrievalError {
    /// Unable to retrieve the locale
    NullLocale,
    ...
    Windows(windows_result::Error),
    Other(String),
}
```

Next, you can implement a `From`:

```
impl From<windows_result::Error> for RetrievalError {
    fn from(input: windows_result::Error) -> Self {
        Self::Windows(input)
    }
}
```

Once you have this, in this `get_locales` you can use `?` which will auto convert and propagate the error:

```
let locale = GlobalizationPreferences::Languages()?;

for i in 0.. languages.Size()? {
    let hstring = languages.GetAt(i)?;
    let string = hstring.to_string_lossy();
    locale_vec_str.push(string);
}
```

And once you're here, the iterator combinators are very powerful in Rust (similar to Python) and optimize really well. You start with a [`IVectorView`](https://microsoft.github.io/windows-docs-rs/doc/windows/System/UserProfile/struct.GlobalizationPreferences.html#method.Languages).

It implements a trait called `IntoIterator`, great! So we can iterate over the elements without having to rely on Size and then GetAt pair:

```
let input_locales = GlobalizationPreferences::Languages()?;
let result_locales: Vec<String> = languages
    .into_iter()
    .map(|hstring| { hstring.to_string_lossy() })
    .collect();

Ok(result_locales)
```

ashu26jha replied:
Great take @zbraniecki

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,155 @@
+                            languages.push(lang_rust_str.to_string());
+                        }
+                        Err(e) => {
+                            return Err(RetrievalError::ConversionError(e));
```
you can do the same here - add `From<Utf8Error> for RetrievalError` and then you can use `?` to propagate errors.

ashu26jha replied:
Makes sense now to ensure consistency 👍

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,155 @@
+            if !locale_cstr.is_null() {
+                // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+                let calendar_rust_str_res = unsafe { CStr::from_ptr(locale_cstr) }.to_str();
+                match calendar_rust_str_res {
```
and here.

zbraniecki commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,155 @@
+                // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+                let identifier_str_res = unsafe { CStr::from_ptr(identifier_cstr) }.to_str();
+                match identifier_str_res {
+                    Ok(identifier_str) => {
```
and here.

zbraniecki commented on utils/env_preferences/src/linux.rs:
```diff
@@ -0,0 +1,111 @@
+
+    if !locale_ptr.is_null() {
+        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+        let c_str = unsafe { CStr::from_ptr(locale_ptr) };
```
and here :)

zbraniecki reviewed (approved):
lgtm! Congrats!

ashu26jha commented:
List of changes made:

1. Converted return of APIs to `Result`
2. Added tests for macos, linux and windows
3. Added `Time Zone` support for macos and windows

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -2,135 +2,203 @@
+    // that the `lang_ptr` is not NULL thus making it safe to call
+    let length = unsafe { CFStringGetLength(ptr) as usize };
+
+    let mut c_str_buf = vec![0; length * 4];
```
issue: why is this `length * 4`? I think this should be `length + 1`: https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring#parameters

ashu26jha replied:
*4 is an overkill 😅

robertbastian replied:
it's underkill and maybe UB if the string is empty.

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -2,135 +2,203 @@
+        );
+    }
+
+    let c_str_buf_u8: Vec<u8> = c_str_buf.iter().map(|&c| c as u8).collect();
```
this might or might not optimise out. It's probably better to instantiate `c_str_buf_u8` as `Vec<u8>`, and do a pointer cast to `*mut c_char` when passing it to `CFStringGetCString`

ashu26jha replied:
Not really sure how to do this tbh. `c_str_buf_u8 ` is already an `Vec<u8>`.

If I am not mistaken are you wishing to have `c_str_buf_u8: Vec<u8>` be the buffer instead of `c_str_buf`?

ashu26jha replied:
I think this is better no point in having two allocations

robertbastian replied:
Sorry I meant instantiate `c_str_buf` as `Vec<u8>`. Currently that gets inferred to `Vec<c_char>`, and this `.collect` reallocates.

robertbastian replied:
ah you did this

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -2,135 +2,203 @@
+    }
+
+    let c_str_buf_u8: Vec<u8> = c_str_buf.iter().map(|&c| c as u8).collect();
+    let str_converted = String::from_utf8_lossy(&c_str_buf_u8);
```
use `CString::from_vec_with_nul` and `CString::into_string`

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -2,135 +2,203 @@
+
+use crate::RetrievalError;
+
+// Helper function
```
I'm confused by what this actually does. In other places you do

```rust
CStr::from_ptr(
    CFStringGetCStringPtr(
        ptr as CFStringRef, 
        kCFStringEncodingUTF8
    )
)
.to_str()?
.to_string()
```

which seems to do the same without manual allocations.

ashu26jha replied:
Actually, this is a first priority to do without manual allocations. I had test failing in CI (not locally). Later realised this on apple [website](https://developer.apple.com/documentation/corefoundation/1542133-cfstringgetcstringptr?language=objc):

```
This function either returns the requested pointer immediately, 
with no memory allocations and no copying, 
in constant time, or returns NULL. If the latter is the result, call 
an alternative function such as the 
CFStringGetCString function to extract the characters.
```

robertbastian replied:
Please encapsulate the whole logic, including the null check in this function then, and add the comment explaining what you're doing. `get_string` doesn't tell me anything.

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,212 @@
+/// Helps to get string, when a string cannot be obtained in O(1) time and space directly from the pointer
+/// A buffer is created of size `length + 1`. After copying the string into the buffer, the null characters `\0`
+/// are removed from the string, which helps it converting to icu4x locale in tests
+fn get_string(ptr: CFStringRef) -> Result<String, RetrievalError> {
```
please put the whole logic with both branches of the if block in this method. The case where you go through`CStr` is duplicated multiple times below, and should be part of `get_string`.

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -22,31 +22,44 @@ use crate::RetrievalError;
+    // SAFETY: The call to `CFStringGetCStringPtr` because the reference of string we are accessing is not `NULL`
+    // Returns pointer in O(1) without any memory allocation. This can return NULL so we are handling it by directly
+    // copying it using `CFStringGetCString`
+    let lang_str: *const c_char = unsafe { CFStringGetCStringPtr(ptr, kCFStringEncodingUTF8) };
```
nit: it's not `lang_` here

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,167 @@
+
+        let c_string = CString::from_vec_with_nul(c_str_buf)?;
+        let str_converted = c_string.into_string()?;
+        Ok(str_converted
```
`CString::into_string` doesn't include the trailing `NUL`, so you shouldn't do this (it's an extra allocation).

robertbastian commented on utils/env_preferences/src/error.rs:
```diff
@@ -0,0 +1,66 @@
+#[derive(Debug)]
+pub enum RetrievalError {
+    /// Error converting from `CString` to `String`
+    CStringConversionError(IntoStringError),
```
`IntoStringError` contains the `CString` and a `Utf8Error`. We only care about the UTF-8 error, we don't want to propagate the invalid `CString` to the caller.

robertbastian commented on utils/env_preferences/src/apple.rs:
```diff
@@ -0,0 +1,167 @@
+                let locale_str = get_string(lang_ptr as CFStringRef)?;
+                languages.push(locale_str);
+            } else {
+                return Err(RetrievalError::NullPointer);
```
why is this returning `NullPointer` instead of `NullLocale`?

ashu26jha replied:
Because it's a pointer to the stored locale, the reference to `CFString`

robertbastian replied:
Isn't the case where you return `NullLocale` just that, a null string?

ashu26jha replied:
Ok, if I think `NullLocale` is more descriptive as, if the we are getting a null it implies we couldn't the locale from the array

robertbastian replied:
You should now be able to get rid of the `NullPointer` variant.

sffc commented on utils/env_preferences/tests/test.rs:
```diff
@@ -0,0 +1,153 @@
+        for locale in locale_res.into_values() {
+            let parts: Vec<&str> = locale.split('.').collect();
+
+            // Skipping "C" and those ending with "UTF-8", as they cannot be converted
+            // into the locale
```
Why can't they be converted into a `Locale`? Can they be converted into something else? Can we check that the strings are in a form that we expect, even if not a BCP-47 Locale?

ashu26jha replied:
Tried, doing the following:
```rust
let loc_str = "C";
let locale_ loc.parse.unwrap();
```
It fails

robertbastian replied:
Converting to locale is a whole other can of worms, there will be strings like `C`, `de_CH.UTF-8`, etc. Let's not scope-creep this PR and stick with strings for now.

sffc replied:
I think I've been clear about the line I draw given that this is unsafe code:

1. All code paths must be covered
2. All bytes resulting from calls to unsafe functions must be read
3. The _shape_ of the output should be asserted

I am relying on @robertbastian and @zbraniecki to verify condition 1 (coverage).

I guess condition 2 is covered in `std::ffi::CStr::to_str`.

For condition 3, if the strings are not valid BCP-47, then let's assert what shape they are. The bare minimum would be an assertion that they are ASCII, for example. But condition 3 wouldn't rise to the level of blocking the PR from landing in experimental.

sffc commented on utils/env_preferences/tests/test.rs:
```diff
@@ -0,0 +1,153 @@
+    #[test]
+    fn test_calendar() {
+        let calendar_res = get_system_calendars().unwrap();
+        for calendar in calendar_res {
+            assert!(!calendar.0.is_empty(), "Couldn't retreive calendar locale");
+            assert!(!calendar.1.is_empty(), "Couldn't retreive calendar");
+        }
```
Nit: Testing nonempty is not sufficient. Please actually read the strings. Assert _something_ about their shape. Maybe have a list of valid values? At the very least check that they are ASCII-alphabetic, for example.

ashu26jha replied:
The issue with a list is, we are not sure about the people who are contributing to the icu4x, they could have some different locale and it would display failing test. 

Check for ascii can be done

sffc commented:
I removed my blocking review; I will let @robertbastian or @zbraniecki land the PR.
